wasm/jit/jit_amd64_test.go

```diff
@@ -0,0 +1,1441 @@
+	compiledFunc.codeInitialAddress = uintptr(unsafe.Pointer(&compiledFunc.codeSegment[0]))
+	eng.compiledWasmFunctions = []*compiledWasmFunction{compiledFunc}
+	// Call into the function
+	eng.exec(compiledFunc)
+	require.Equal(t, []uint64{5, 0, 0, 0, 0}, eng.stack[:5])
+	// And the callstack should be empty.
+	require.Nil(t, eng.callFrameStack)
+
+	// // Check the stability with busy Go runtime.
+	var wg sync.WaitGroup
+	const goroutines = 10000
+	wg.Add(goroutines)
+	for i := 0; i < goroutines; i++ {
+		i := i
+		go func() {
+			if i/10 == 0 {
+				// This is to kick the Go runtime to come in
+				fmt.Fprintf(io.Discard, "aaaaaaaaaaaa")
+			}
+			defer wg.Done()
+			// Setup engine.
+			mem := newMemoryInst()
+			eng := newEngine()
+			eng.stack[0] = 10 // We call recursively 10 times.
+			compiledFunc := &compiledWasmFunction{codeSegment: code, memory: mem, inputs: 1, returns: 1}
+			compiledFunc.codeInitialAddress = uintptr(unsafe.Pointer(&compiledFunc.codeSegment[0]))
+			eng.compiledWasmFunctions = []*compiledWasmFunction{compiledFunc}
+			// Call into the function
+			eng.exec(compiledFunc)
+		}()
+	}
+}
+
+// Test perform operations on
+// pushing the const value into the Go-allocated slice
+// under large amout of Goroutines.
+func TestPushValueWithGoroutines(t *testing.T) {
+	const (
+		targetValue        uint64 = 100
+		pushTargetRegister        = x86.REG_AX
+		goroutines                = 10000
+	)
+	var wg sync.WaitGroup
+	wg.Add(goroutines)
+	for i := 0; i < goroutines; i++ {
+		go func() {
+			defer wg.Done()
+			// Build native codes.
+			builder := requireNewBuilder(t)
+			builder.initializeReservedRegisters()
+			// Push consts to pushTargetRegister.
+			_ = builder.locationStack.pushValueOnStack() // dummy value, not actually used!
+			loc := builder.locationStack.pushValueOnRegister(pushTargetRegister)
+			builder.movConstToRegister(int64(targetValue), pushTargetRegister)
+			// Push pushTargetRegister into the engine.stack[engine.sp].
+			builder.releaseRegister(loc)
+			// Finally increment the stack pointer and write it back to the eng.sp
+			builder.returnFunction()
+
+			// Compile.
+			code, err := builder.assemble()
+			require.NoError(t, err)
+
+			eng := newEngine()
+			mem := newMemoryInst()
+
+			f := &compiledWasmFunction{codeSegment: code, memory: mem}
+			f.codeInitialAddress = uintptr(unsafe.Pointer(&f.codeSegment[0]))
+
+			// Call into the function
+			eng.exec(f)
+
+			// Because we pushed the value, eng.sp must be incremented by 1
+			if eng.currentStackPointer != 2 {
+				panic("eng.sp must be incremented.")
+			}
+
+			// Also we push the const value to the top of slice!
+			if eng.stack[1] != 100 {
+				panic("eng.stack[0] must be changed to the const!")
+			}
+		}()
+	}
+	wg.Wait()
+}
+
+func Test_setJITStatus(t *testing.T) {
+	for _, s := range []jitCallStatusCode{jitCallStatusCodeReturned, jitCallStatusCodeCallWasmFunction, jitCallStatusCodeCallBuiltInFunction} {
```

Comment:
should this  and below be doing t.Run?

Reply:
https://github.com/tetratelabs/wazero/pull/60/commits/5fd13b36250f0fb937bfc3867b52f4848d7f0022